include/boost/geometry/algorithms/detail/overlay/approximately_equals.hpp

```diff
@@ -0,0 +1,53 @@
+// Boost.Geometry (aka GGL, Generic Geometry Library)
+
+// Copyright (c) 2021 Barend Gehrels, Amsterdam, the Netherlands.
+
+// Use, modification and distribution is subject to the Boost Software License,
+// Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_GEOMETRY_ALGORITHMS_DETAIL_OVERLAY_APPROXIMATELY_EQUALS_HPP
+#define BOOST_GEOMETRY_ALGORITHMS_DETAIL_OVERLAY_APPROXIMATELY_EQUALS_HPP
+
+#include <boost/geometry/util/math.hpp>
+#include <boost/geometry/util/select_coordinate_type.hpp>
+#include <boost/geometry/util/select_most_precise.hpp>
+
+namespace boost { namespace geometry
+{
+
+#ifndef DOXYGEN_NO_DETAIL
+namespace detail { namespace overlay
+{
+
+template <typename Point1, typename Point2, typename E>
+inline bool approximately_equals(Point1 const& a, Point2 const& b,
+                                        E const& limit_kilo_epsilon)
+{
+    // Including distance would introduce cyclic dependencies.
+    using coor_t = typename select_coordinate_type<Point1, Point2>::type;
+    using calc_t = typename geometry::select_most_precise<coor_t, E>::type;
+    constexpr calc_t machine_kilo_epsilon = 1.0e3 * std::numeric_limits<calc_t>::epsilon();
+
+    calc_t const& a0 = geometry::get<0>(a);
+    calc_t const& b0 = geometry::get<0>(b);
+    calc_t const& a1 = geometry::get<1>(a);
+    calc_t const& b1 = geometry::get<1>(b);
+    calc_t const one = 1.0;
+    calc_t const c = math::detail::greatest(std::abs(a0), std::abs(b0),
+                                            std::abs(a1), std::abs(b1), one);
+
+    // The maximum limit is avoid, for floating point, large limits like 400
+    // (which are be calculated using eps)
+    constexpr calc_t maxlimit = 1.0e-3;
+    auto const limit = (std::min)(maxlimit, limit_kilo_epsilon * machine_kilo_epsilon * c);
+    return std::abs(a0 - b0) <= limit && std::abs(a1 - b1) <= limit;
```

Comment:
The comment suggests that an alternative for this algorithm is the one using `distance`. `distance` would take all of the dimensions into account. This algorithm takes only 2 dimensions into account. Which one is correct? In general case should this be replaced with a N-dimensional version?

----

In `namespace math` there are:
- `equals_by_policy` - https://github.com/boostorg/geometry/blob/develop/include/boost/geometry/util/math.hpp#L196
- `equals_with_epsilon` - https://github.com/boostorg/geometry/blob/develop/include/boost/geometry/util/math.hpp#L575

Maybe they could be used here instead of calculating all of this manually?
If needed the policy in `equals_by_policy` could be replaced with lambda expression for convenience.

Reply:
Furthermore, it would be possible to use distance here. You could simply pass Pt/Pt distane strategy or newly proposed umbrella strategy (https://github.com/boostorg/geometry/pull/812) and then use it directly here without including the algorithm.

But have in mind that this computionally simple code would become a lot more complex if distance was calculated. There would be additional `sqrt` call for cartesian and `sin/cos` calls for other coordinate systems.

Reply:
I moved this function. It's not new in this PR.
But yes, I recently created it and it could be enhanced or replaced.
The problem with the code from math is that it gives too high values for floating point (as indicated in the comment).
I tried to replace it with comparable distance, and that works for cartesian but (again) requires an epsilon defined by the coordinate system, which will result in high values for float and therefore in similar code as is written here.
The 3d variant we could make with your new `for_each_coordinate` functionality.

For now I keep it here but we can discuss further about it.

Reply:
I did a test using `distance`. W.r.t. performance in Cartesian there is no measurable difference in this use case. Probably because it doesn't happen so often.
But using geographical distance will cause problems:
```
false 1.6948370905798346708337476229644
false 0.84474950387190805844994656581548
false 0.41427736018131078310489101568237
false 0.19008348922390885804567517425312
false 0
false 0
```
From the small test, the distance already returns zero if they are still far away (in terms of epsilon). Apart from that it's probably an overkill to use geographic distance there.

So I propose to still keep it as it is here (and using `for_each` later)

Reply:
`equals_with_epsilon` and `equals_with_policy` allows you to pass a value which is multiplied by machine epsilon before performing the test. Just like you did.

Reply:
I'll have a look again.

Reply:
:heavy_check_mark: thanks for the comments, done in new commit. I added a unit test too.
_(I will squash commits before merging)_